### PR TITLE
Update crawl.py - for loops to list comprehension

### PR DESCRIPTION
**Pull request by RahulDubey391**
Eliminated nested for loops and replaced it with list comprehension.

### CODE REVIEW

**Comment by RahulDubey391** [comment]
@mshudrak Hi, can you tell me if this is okay for merging? Please leave any feedback so I can work it up. Thanks!!!

File: src/gcp_scanner/crawl.py [hunk 1]
```diff
@@ -125,9 +124,7 @@ def get_compute_instances_names(
     while request is not None:
       response = request.execute()
       if response.get("items", None) is not None:
-        for _, instances_scoped_list in response["items"].items():
-          for instance in instances_scoped_list.get("instances", []):
-            images_result.append(instance)
+        images_result = [instance for _, instances_scoped_list in response["items"].items() for instance in instances_scoped_list.get("instances",[])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:127]: Split to multiple lines for readability.

File: src/gcp_scanner/crawl.py [hunk 2]
```diff
@@ -186,9 +182,7 @@ def get_compute_disks_names(
     while request is not None:
       response = request.execute()
       if response.get("items", None) is not None:
-        for _, disks_scoped_list in response["items"].items():
-          for disk in disks_scoped_list.get("disks", []):
-            disk_names_list.append(disk)
+        disk_names_list = [disk for _, disks_scoped_list in response["items"].items() for disk in disks_scoped_list.get("disks", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:185]: Split to multiple lines for readability.

File: src/gcp_scanner/crawl.py [hunk 3]
```diff
@@ -218,11 +212,8 @@ def get_static_ips(project_name: str,
     request = service.addresses().aggregatedList(project=project_name)
     while request is not None:
       response = request.execute()
-      for name, addresses_scoped_list in response["items"].items():
-        if addresses_scoped_list.get("addresses", None) is None:
-          continue
-        ips_list.append({name: addresses_scoped_list})
-
+      ips_list = [{name: addresses_scoped_list} for name, addresses_scoped_list in response["items"].items() if addresses_scoped_list.get("addresses", None) is not None]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:215]: Split to multiple lines for readability.

File: src/gcp_scanner/crawl.py [hunk 4]
```diff
@@ -562,9 +543,8 @@ def get_bq(project_id: str,
       response = request.execute()
 
       for dataset in response.get("datasets", []):
-        dataset_id = dataset["datasetReference"]["datasetId"]
-        bq_datasets[dataset_id] = get_bq_tables(project_id, dataset_id, service)
-
+        bq_datasets[dataset["datasetReference"]["datasetId"]] = get_bq_tables(project_id, dataset["datasetReference"]["datasetId"], service)
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:546]: Removing the temporary variable decreases readability, revert.

File: src/gcp_scanner/crawl.py [hunk 5]
```diff
@@ -961,11 +934,7 @@ def get_service_accounts(project_name: str,
     request = service.projects().serviceAccounts().list(name=name)
     while request is not None:
       response = request.execute()
-
-      for service_account in response.get("accounts", []):
-        service_accounts.append(
-            (service_account["email"], service_account["description"]
-             if "description" in service_account else ""))
+      service_accounts = [(service_account["email"], service_account["description"] if "description" in service_account else "") for service_account in response.get("accounts", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:937]: Split to multiple lines for readability.

**Review by ZetaTwo — CHANGES_REQUESTED** [review]

File: src/gcp_scanner/crawl.py (hunk 1 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: Hi @ZetaTwo , I have resolved this issue in the latest PR.

File: src/gcp_scanner/crawl.py (hunk 3 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: Hi @ZetaTwo , I have resolved this issue in the latest PR.

File: src/gcp_scanner/crawl.py (hunk 5 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: Hi @ZetaTwo , I have resolved this issue in the latest PR.

File: src/gcp_scanner/crawl.py (hunk 2 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: Hi @ZetaTwo , I have resolved this issue in the latest PR.

File: src/gcp_scanner/crawl.py (hunk 4 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: Hi @ZetaTwo , I have resolved this issue in the latest PR.

File: src/gcp_scanner/crawl.py [hunk 6]
```diff
@@ -281,8 +280,8 @@ def get_subnets(project_name: str,
     while request is not None:
       response = request.execute()
       if response.get("items", None) is not None:
-        for name, subnetworks_scoped_list in response["items"].items():
-          subnets_list.append((name, subnetworks_scoped_list))
+        subnets_list = [(name, subnetworks_scoped_list) 
+                        for name, subnetworks_scoped_list in response["items"].items()]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:284]: This is redundant. Replace with just `subnets_list = response["items"].items()`

File: src/gcp_scanner/crawl.py [hunk 7]
```diff
@@ -495,8 +491,8 @@ def get_sql_instances(project_name: str,
     request = service.instances().list(project=project_name)
     while request is not None:
       response = request.execute()
-      for database_instance in response.get("items", []):
-        sql_instances_list.append(database_instance)
+      sql_instances_list = [database_instance 
+                            for database_instance in response.get("items", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:495]: This is redundant. Replace with `sql_instances_list = response.get("items", [])`

File: src/gcp_scanner/crawl.py [hunk 8]
```diff
@@ -528,8 +524,8 @@ def get_bq_tables(project_id: str, dataset_id: str,
     while request is not None:
       response = request.execute()
 
-      for table in response.get("tables", []):
-        list_of_tables.append(table)
+      list_of_tables = [table 
+                        for table in response.get("tables", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:528]: This is redundant. Replace with `list_of_tables = response.get("tables", [])`

File: src/gcp_scanner/crawl.py [hunk 9]
```diff
@@ -562,9 +558,9 @@ def get_bq(project_id: str,
       response = request.execute()
 
       for dataset in response.get("datasets", []):
-        dataset_id = dataset["datasetReference"]["datasetId"]
-        bq_datasets[dataset_id] = get_bq_tables(project_id, dataset_id, service)
-
+        datasetid = dataset["datasetReference"]["datasetId"]
+        bq_datasets[datasetid] = get_bq_tables(project_id,datasetid, service)
+
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:563]: Trim trailing whitespace

File: src/gcp_scanner/crawl.py [hunk 10]
```diff
@@ -595,9 +591,9 @@ def get_pubsub_subscriptions(project_id: str,
         project=f"projects/{project_id}")
     while request is not None:
       response = request.execute()
-      for subscription in response.get("subscriptions", []):
-        pubsubs_list.append(subscription)
-
+      pubsubs_list = [subscription 
+                      for subscription in response.get("subscriptions", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:595]: This is redundant. Replace with `pubsubs_list = response.get("subscriptions", [])`

File: src/gcp_scanner/crawl.py [hunk 11]
```diff
@@ -595,9 +591,9 @@ def get_pubsub_subscriptions(project_id: str,
...
+
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:596]: Trim trailing whitespace

File: src/gcp_scanner/crawl.py [hunk 12]
```diff
@@ -627,8 +623,8 @@ def get_cloudfunctions(project_id: str,
         parent=f"projects/{project_id}/locations/-")
     while request is not None:
       response = request.execute()
-      for function in response.get("functions", []):
-        functions_list.append(function)
+      functions_list = [function 
+                        for function in response.get("functions", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:627]: This is redundant. Replace with `functions_list = response.get("functions", [])`

File: src/gcp_scanner/crawl.py [hunk 13]
```diff
@@ -661,8 +657,8 @@ def get_bigtable_instances(project_id: str,
         parent=f"projects/{project_id}")
     while request is not None:
       response = request.execute()
-      for instance in response.get("instances", []):
-        bigtable_instances_list.append(instance)
+      bigtable_instances_list = [instance 
+                                 for instance in response.get("instances", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:661]: This is redundant. Replace with `bigtable_instances_list = response.get("instances", [])`

File: src/gcp_scanner/crawl.py [hunk 14]
```diff
@@ -695,9 +691,9 @@ def get_spanner_instances(project_id: str,
         parent=f"projects/{project_id}")
     while request is not None:
       response = request.execute()
-      for instance in response.get("instances", []):
-        spanner_instances_list.append(instance)
-
+      spanner_instances_list = [instance 
+                                for instance in response.get("instances", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:695]: This is redundant. Replace with `spanner_instances_list = response.get("instances", [])`

File: src/gcp_scanner/crawl.py [hunk 15]
```diff
@@ -695,9 +691,9 @@ def get_spanner_instances(project_id: str,
...
+
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:696]: Trim trailing whitespace

File: src/gcp_scanner/crawl.py [hunk 16]
```diff
@@ -728,8 +724,8 @@ def get_filestore_instances(project_id: str,
         parent=f"projects/{project_id}/locations/-")
     while request is not None:
       response = request.execute()
-      for instance in response.get("instances", []):
-        filestore_instances_list.append(instance)
+      filestore_instances_list = [instance 
+                                  for instance in response.get("instances", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:728]: Redundant.

File: src/gcp_scanner/crawl.py [hunk 17]
```diff
@@ -822,8 +818,8 @@ def get_app_services(project_name: str,
     app_services["services"] = list()
     while request is not None:
       response = request.execute()
-      for service_entry in response.get("services", []):
-        app_services["services"].append(service_entry)
+      app_services['services'] = [service_entry 
+                                  for service_entry in response.get("services", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:822]: Redundant.

File: src/gcp_scanner/crawl.py [hunk 18]
```diff
@@ -857,9 +853,9 @@ def get_endpoints(project_id: str,
     request = service.services().list(producerProjectId=project_id)
     while request is not None:
       response = request.execute()
-      for service_entry in response.get("services", []):
-        endpoints_list.append(service_entry)
-
+      endpoints_list = [service_entry 
+                        for service_entry in response.get("services", [])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:857]: Redundant.

File: src/gcp_scanner/crawl.py [hunk 19]
```diff
@@ -857,9 +853,9 @@ def get_endpoints(project_id: str,
...
+
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:858]: Whitespace

File: src/gcp_scanner/crawl.py [hunk 20]
```diff
@@ -95,8 +95,8 @@ def get_project_list(credentials: Credentials) -> List[Dict[str, Any]]:
     while request is not None:
       response = request.execute()
 
-      for project in response.get("projects", []):
-        project_list.append(project)
+      project_list = [project 
+                      for project in response.get("projects",[])]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:99]: Redundant.

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
Several of the list comprehensions are on the form `x = [y for y in z]` and can be replaced with just `x = z`, unless there is a specific reason to not do this.

File: src/gcp_scanner/crawl.py (hunk 6 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py:274]: Hi @ZetaTwo, I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 7 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 8 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 9 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 10 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 11 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 12 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 13 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 14 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 15 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 16 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 17 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py:792]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 18 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 19 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 20 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: I have made the requested changes in the latest PR. Check and let me know if any changes required.

File: src/gcp_scanner/crawl.py (hunk 6 above)
Comment (ZetaTwo) [reply src/gcp_scanner/crawl.py:274]: This is still not fixed

File: src/gcp_scanner/crawl.py [hunk 21]
```diff
@@ -312,12 +300,8 @@ def get_firewall_rules(
     request = compute_client.firewalls().list(project=project_name)
     while request is not None:
       response = request.execute()
-      for firewall in response.get("items", []):
-        firewall_rules_list.append((
-            firewall["name"],
-            # firewall['description'],
-        ))
-
+      firewall_rules_list = [(firewall['name'])
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:303]: missing comma in the tuple

File: src/gcp_scanner/crawl.py [hunk 22]
```diff
@@ -562,9 +539,8 @@ def get_bq(project_id: str,
       response = request.execute()
 
       for dataset in response.get("datasets", []):
-        dataset_id = dataset["datasetReference"]["datasetId"]
-        bq_datasets[dataset_id] = get_bq_tables(project_id, dataset_id, service)
-
+        datasetid = dataset["datasetReference"]["datasetId"]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:542]: keep the snake_case

File: src/gcp_scanner/crawl.py (hunk 17 above)
Comment (ZetaTwo) [reply src/gcp_scanner/crawl.py:792]: This is still not fixed.

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
Looks pretty good overall but I left some more comments.

**Comment by RahulDubey391** [comment]
@ZetaTwo Hi, I have applied the missing changes, also update me if I have missed anything else.

**Comment by RahulDubey391** [comment]
@mshudrak , @ZetaTwo  Hi, it seems there are cases of long-lines and trailing-spaces, I'll work on it to eliminate the issues.

**Comment by RahulDubey391** [comment]
@ZetaTwo @mshudrak Hi, The latest PR has passed all the checks. Can you please review and let me know if there are changes still need to be made?

File: src/gcp_scanner/crawl.py [hunk 23]
```diff
@@ -312,12 +299,8 @@ def get_firewall_rules(
     request = compute_client.firewalls().list(project=project_name)
     while request is not None:
       response = request.execute()
-      for firewall in response.get("items", []):
-        firewall_rules_list.append((
-            firewall["name"],
-            # firewall['description'],
-        ))
-
+      firewall_rules_list=list(map(lambda x:(x["name"],),
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:302]: Why did you change this from the list comprehension to this? This is less readable. You were only missing a comma in the previous version.

File: src/gcp_scanner/crawl.py [hunk 24]
```diff
@@ -434,8 +414,7 @@ def get_gke_clusters(
   parent = f"projects/{project_name}/locations/-"
   try:
     clusters = gke_client.list_clusters(parent=parent)
-    return [(cluster.name, cluster.description) for cluster in clusters.clusters
-           ]
+    return list(map(lambda x: (x.name,x.description),clusters.clusters))
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:417]: Why did you change this from the list comprehension to this? This is less readable.

File: src/gcp_scanner/crawl.py [hunk 25]
```diff
@@ -403,10 +386,7 @@ def get_managed_zones(project_name: str,
     request = service.managedZones().list(project=project_name)
     while request is not None:
       response = request.execute()
-
-      for managed_zone in response["managedZones"]:
-        zones_list.append(managed_zone)
-
+      zones_list = response["managedZones"]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:389]: Use `.get(...)` to avoid exceptions just like in the other places

File: src/gcp_scanner/crawl.py [hunk 26]
```diff
@@ -563,8 +537,7 @@ def get_bq(project_id: str,
 
       for dataset in response.get("datasets", []):
         dataset_id = dataset["datasetReference"]["datasetId"]
-        bq_datasets[dataset_id] = get_bq_tables(project_id, dataset_id, service)
-
+        bq_datasets[dataset_id]=get_bq_tables(project_id,dataset_id, service)
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:540]: Keep the whitespace

File: src/gcp_scanner/crawl.py [hunk 27]
```diff
@@ -961,12 +920,8 @@ def get_service_accounts(project_name: str,
     request = service.projects().serviceAccounts().list(name=name)
     while request is not None:
       response = request.execute()
-
-      for service_account in response.get("accounts", []):
-        service_accounts.append(
-            (service_account["email"], service_account["description"]
-             if "description" in service_account else ""))
-
+      service_accounts=list(map(lambda x:(x["email"],x["description"]
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:923]: Why did you change this from the list comprehension to this? This is less readable.

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
You seem to have changed several list comprehensions to list(map(...)). This makes it less readable.

File: src/gcp_scanner/crawl.py (hunk 23 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: reverted the map() to list comprehension

File: src/gcp_scanner/crawl.py (hunk 24 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: reverted the map() to list comprehension

File: src/gcp_scanner/crawl.py (hunk 25 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: used response.get("managedZones",[]) instead of directly referencing the key.

File: src/gcp_scanner/crawl.py (hunk 26 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py:541]: added the white-space

File: src/gcp_scanner/crawl.py (hunk 27 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: reverted the map() to list comprehension

File: src/gcp_scanner/crawl.py [hunk 28]
```diff
@@ -961,12 +922,9 @@ def get_service_accounts(project_name: str,
...
     while request is not None:
       response = request.execute()
-
-      for service_account in response.get("accounts", []):
-        service_accounts.append(
-            (service_account["email"], service_account["description"]
-             if "description" in service_account else ""))
-
+      service_accounts=[(service_account["email"],service_account["description"]
+        if "description" in service_account else "")
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:926]: replace this `if ... else ...` with `.get("description", "")`

**Review by ZetaTwo — CHANGES_REQUESTED** [review]

File: src/gcp_scanner/crawl.py (hunk 28 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: replaced it with .get(key,"")

**Comment by RahulDubey391** [comment]
@ZetaTwo Hi, I have made the requested change. Let me know if anything else needs to be changed. Thanks!!

**Review by ZetaTwo — APPROVED** [review]

File: src/gcp_scanner/crawl.py (hunk 26 above)
Comment (mshudrak) [reply src/gcp_scanner/crawl.py:541]: @RahulDubey391 Are you sure this is resolved?

File: src/gcp_scanner/crawl.py [hunk 29]
```diff
@@ -961,12 +922,9 @@ def get_service_accounts(project_name: str,
     request = service.projects().serviceAccounts().list(name=name)
     while request is not None:
       response = request.execute()
-
-      for service_account in response.get("accounts", []):
-        service_accounts.append(
-            (service_account["email"], service_account["description"]
-             if "description" in service_account else ""))
-
+      service_accounts=[(service_account["email"],
```
Comment (mshudrak) [inline src/gcp_scanner/crawl.py:925]: Whitespaces before and after `=`

**Review by mshudrak — CHANGES_REQUESTED** [review]
There are two small nits to fix but overall almost ready to merge.

File: src/gcp_scanner/crawl.py (hunk 29 above)
Comment (RahulDubey391) [reply src/gcp_scanner/crawl.py]: @mshudrak Hi, I have made the requested change. For '='  space is added before and after. But for extra space between list comprehension and 'request = ' line we kept it so that readability doesn't become an issue. Also for service_account list comprehension, it is split to multiple lines since lint check was causing the issue.

**Review by mshudrak — APPROVED** [review]
Thanks, merging.
